rust/private/rustc.bzl

```diff
@@ -310,15 +316,6 @@ def get_linker_and_args(ctx, cc_toolchain, feature_configuration, rpaths):
 
     return ld, link_args, link_env
 
-def _expand_locations(ctx, env, aspect):
```

Comment:
@illicitonion @dae Do you know why is this different from the macro that was in`cargo/cargo_build_script.bzl`?

Reply:
- The one in rustc is concerned with the compilation stage (rustc_env), so it needs to work when aspects like clippy are enabled. process_wrapper does not support ${pwd} expansion in environmental variables at the moment, so execpath/location will not work - only an unadorned $(rootpath ...) will work
- The one in cargo applies to the environment passed in when running the build script (build_script_env). $(rootpath ...) won't work here because the working directory is not execroot, and the build script runner has been modified to replace ${pwd} references with execroot so that the execroot-relative paths work correctly.
- env is a string possibly containing $(location/execpath/rootpath) references
- data is an attr list of tools from the calling rule that will be used to expand the location references into actual paths

Reply:
Is there a way to unify the functionality? Or do you feel it's more correct to have these separate functions exist?

Reply:
Because of the differences I'm not sure using a single routine is a big win, as you'd need to document the two different use cases it's trying to cover. It may be possible to get location/execpath to work with rustc_env in the future, but that would require modifying process_wrapper, and my C++ is rusty at best.

Reply:
I can update my PR to restore the macros but could you write me some documentation to copy and paste for them? The documentation was really the only thing that bugged me but then thought there might be an easy win to deduplicate code.

Reply:
@dae Ok, I've restored the code but made one small change. Can you give me some text to populate the template with?

Reply:
The text here should be identical to the one I suggested above, but without the caveat about absolute paths (because `rustc` _does_ run in the exec root, so exec-root-relative paths are fine)

Reply:
Thanks friends! 😄 🙏